LTD: Add model to track case queue movements

saruniitr opened the pull request:
## Change description

This model is used to track the queues the case been through. We create instance of this model when a case is assigned to a particular queue.

Back populate data by going through MOVE_CASE events.
The back population of `exit_date` is not done for the reasons mentioned in the comments. I am looking for comments/suggestions to help with this.

Further changes required in `lite_routing` and API after these changes are merged (https://github.com/uktrade/lite-routing/pull/310). As lite_routing is a submodule, I need these changes merged first to get access to the model there.

lite_routing PR is merged so sha is updated in this PR.


[LTD-5866]: https://uktrade.atlassian.net/browse/LTD-5866?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kevincarrogan commented on api/cases/migrations/0077_populate_case_queue_movements.py:
```diff
@@ -0,0 +1,43 @@
+
+    case_queue_movements = []
+
+    # Some of the early events doesn't have queue_ids in the payload so they are ignored
```
They did have queue names so in theory we could use that data although do you think it's possibly not worth it?

saruniitr replied:
yes, as it is very old case I think its not worth otherwise more queries in the migration.

kevincarrogan replied:
I think the problem might be that I'm going to use this model for Data Workspace as well so I think that we might need that historical data.

saruniitr replied:
ok, I will pick up those events as well

saruniitr replied:
this is updated

kevincarrogan commented on api/cases/migrations/0077_populate_case_queue_movements.py:
```diff
@@ -0,0 +1,43 @@
+        queue_id = event.payload["queue_ids"][0]
+
+        case_queue_movements.append(
+            CaseQueueMovement(
+                case_id=case_id,
+                queue_id=queue_id,
+            )
```
Where are we setting the dates on these objects?

saruniitr replied:
yes, my bad, fixed it.

saruniitr commented on api/cases/models.py:
```diff
@@ -871,3 +871,10 @@ class Meta:
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    case = models.ForeignKey(Case, related_name="casequeuemovements", on_delete=models.DO_NOTHING)
+    queue = models.ForeignKey(Queue, related_name="casequeuemovements", on_delete=models.DO_NOTHING)
+    exit_date = models.DateTimeField(blank=True, null=True)
```
There is no migration added to back populate this field for the following reasons.
1. In some cases the Case went back to the same queue again, it is not possible to correct identify the correct instances and update this field.
2. We need to use `AuditType.UNASSIGNED_QUEUES` and `AuditType.UNASSIGNED` to populate this field. For the second event the queue details are not included in the `payload` so we don't have this data.

I am open to any suggestions to help with (1) to get this field populated.
Instead of populating incorrect data I am thinking it is better to avoid back population.

The idea of this field is we record start and end dates so we can accurately check the time spent on each queue.

currycoder replied:
Something that could be good to record here is `orchestrator`.  e.g. lite routing, a user assigning the queue manually, etc.
Maybe not for right now, though

depsiatwal commented on api/cases/migrations/0077_populate_case_queue_movements.py:
```diff
@@ -0,0 +1,70 @@
+    )
+    for event in move_case_qs:
+        case_id = event.action_object_object_id
+        queue_id = event.payload["queue_ids"][0]
```
have we sanity checked prod to make sure this payload is always there ?

saruniitr replied:
It should be there, will be deploying to `demo` and test

depsiatwal commented on api/cases/migrations/0077_populate_case_queue_movements.py:
```diff
@@ -0,0 +1,70 @@
+                )
+            )
+
+    CaseQueueMovement.objects.bulk_create(case_queue_movements)
```
have we also checked timings or test on PROD like data ?

Just thinking of migration timeouts though it may on be an issue on DBT Platform i'll double check where this migration takes place.

saruniitr replied:
I am going to deploy this to `demo` and check.

kevincarrogan commented on api/cases/migrations/0077_populate_case_queue_movements.py:
```diff
@@ -0,0 +1,70 @@
+    )
+    for event in move_case_qs:
+        case_id = event.action_object_object_id
+        queue_id = event.payload["queue_ids"][0]
```
Why are we only populating the first one of these?

saruniitr replied:
It is only one event for each queue.
https://github.com/uktrade/lite-routing/blob/main/routing_rules_internal/routing_engine.py#L49

kevincarrogan commented on api/cases/migrations/0077_populate_case_queue_movements.py:
```diff
@@ -0,0 +1,70 @@
+    )
+    for event in move_case_qs:
+        case_id = event.action_object_object_id
+        queue_name = event.payload["queues"]
```
I believe this might be an array as well as just a stirng.

saruniitr replied:
queue_name for some reason is not a list, it is not consistent.
eg https://github.com/uktrade/lite-routing/blob/main/routing_rules_internal/routing_engine.py#L49
https://github.com/uktrade/lite-api/blob/dev/api/workflow/user_queue_assignment.py#L66

kevincarrogan replied:
```
In [9]: set([len(a) for a in Audit.objects.filter(verb="move_case").values_list("payload__queue_ids", flat=True) if a])
Out[9]: {1, 2, 3, 5}
```

saruniitr replied:
this is updated

currycoder commented on api/cases/migrations/0077_populate_case_queue_movements.py:
```diff
@@ -0,0 +1,70 @@
+
+
+@transaction.atomic
+def populate_case_queue_movements(apps, schema_editor):
```
Has this migration been run on an environment with representative data?  e.g. one of our demo envs which has anonymised data

saruniitr replied:
I am going to deploy to `demo` and test

saruniitr replied:
deployed to DBT `demo` env, build and deploy successful. model instances created as well.

saruniitr commented:
@kevincarrogan @depsiatwal @currycoder 
Addressed comments, deployed to DBT `demo`, build and deploy successful so migration applied correctly.
I can see instances of the model created as well.
```
In [1]: from api.cases.models import *

In [2]: CaseQueueMovement.objects.count()
Out[2]: 108216

In [3]: from django.conf import settings

In [4]: settings.ENV
Out[4]: 'demo'
```
